traits/tests/test_traits_listener.py

```diff
@@ -0,0 +1,351 @@
+        self.assertEqual(listener.dispatch, "")
+        self.assertTrue(listener.notify)
+        self.assertFalse(listener.is_list_handler)
+        self.assertEqual(listener.type, traits_listener.ANY_LISTENER)
+        self.assertIsNone(listener.next)
+
+    def test_listener_parser_trait_of_trait_of_trait_mixed(self):
+        text = "parent.child1:child2"
+        parser = traits_listener.ListenerParser(text=text)
+
+        # parent listener
+        listener = parser.listener
+        self.assertTrue(listener.notify, "'.' indicates notifications.")
+        self.assertEqual(listener.name, "parent")
+        self.assertEqual(listener.metadata_name, "")
+        self.assertTrue(listener.metadata_defined)
+        self.assertFalse(listener.is_any_trait)
+        self.assertEqual(listener.dispatch, "")
+        self.assertFalse(listener.is_list_handler)
+        self.assertEqual(listener.type, traits_listener.ANY_LISTENER)
+
+        # child1 listener
+        listener = listener.next
+        self.assertFalse(listener.notify, "':' indicates no notifications.")
+        self.assertEqual(listener.name, "child1")
+        self.assertEqual(listener.metadata_name, "")
+        self.assertTrue(listener.metadata_defined)
+        self.assertFalse(listener.is_any_trait)
+        self.assertEqual(listener.dispatch, "")
+        self.assertFalse(listener.is_list_handler)
+        self.assertEqual(listener.type, traits_listener.ANY_LISTENER)
+
+        # child2 listener
+        listener = listener.next
+        self.assertEqual(listener.name, "child2")
+        self.assertEqual(listener.metadata_name, "")
+        self.assertTrue(listener.metadata_defined)
+        self.assertFalse(listener.is_any_trait)
+        self.assertEqual(listener.dispatch, "")
+        self.assertTrue(listener.notify)
+        self.assertFalse(listener.is_list_handler)
+        self.assertEqual(listener.type, traits_listener.ANY_LISTENER)
+        self.assertIsNone(listener.next)
+
+    def test_parse_comma_separated_text(self):
+        text = "child1, child2, child3"
+        parser = traits_listener.ListenerParser(text=text)
+
+        listener_group = parser.listener
+
+        self.assertEqual(len(listener_group.items), 3)
+
+        for listener in listener_group.items:
+            self.assertEqual(listener.metadata_name, "")
+            self.assertTrue(listener.metadata_defined)
+            self.assertFalse(listener.is_any_trait)
+            self.assertEqual(listener.dispatch, "")
+            self.assertTrue(listener.notify)
+            self.assertFalse(listener.is_list_handler)
+            self.assertEqual(listener.type, traits_listener.ANY_LISTENER)
+            self.assertIsNone(listener.next)
+
+        actual = [item.name for item in listener_group.items]
+        self.assertEqual(actual, ["child1", "child2", "child3"])
+
+    def test_parse_comma_separated_text_trailing_comma(self):
+        text = "child1, child2, child3,"
```

Comment:
This one's probably worth a comment and a link to the related issue #406; we want to make this illegal.